packages/opentelemetry-basic-tracer/src/BasicTracer.ts

```diff
@@ -0,0 +1,156 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as types from '@opentelemetry/types';
+import { BaseScopeManager } from '@opentelemetry/scope-base';
+import {
+  ALWAYS_SAMPLER,
+  NoopLogger,
+  NoopSpan,
+  randomSpanId,
+  randomTraceId,
+} from '@opentelemetry/core';
+import { BasicTracerConfig } from '../src/types';
+
+/**
+ * This class represents a basic tracer.
+ */
+export class BasicTracer implements types.Tracer {
+  static INVALID_ID = '0';
+
+  private defaultAttributes: types.Attributes;
+  private logger: types.Logger;
+  private sampler: types.Sampler;
+  private scopeManager?: BaseScopeManager;
+
+  // TODO: consume invalid span context from `SpanContext.INVALID`
+  static defaultSpan = new NoopSpan({
+    traceId: BasicTracer.INVALID_ID,
+    // TODO: consume invalid `spanId` from `Span.INVALID`
+    spanId: '0',
+    traceOptions: types.TraceOptions.UNSAMPLED,
+  });
+
+  /**
+   * Constructs a new Tracer instance.
+   */
+  constructor(config: BasicTracerConfig = {}) {
+    this.sampler = config.sampler || ALWAYS_SAMPLER;
+    this.scopeManager = config.scopeManager;
+    this.defaultAttributes = config.defaultAttributes || {};
+    this.logger = config.logger || new NoopLogger();
+  }
+
+  /**
+   * Starts a new Span or returns the default NoopSpan based on the sampling
+   * decision.
+   */
+  startSpan(name: string, options: types.SpanOptions = {}): types.Span {
+    let parentSpanContext: types.SpanContext | undefined;
+
+    // parent is a SpanContext
+    if (
+      options.parent &&
+      typeof (options.parent as types.SpanContext).traceId
+    ) {
+      parentSpanContext = options.parent as types.SpanContext;
+    }
+    // parent is a Span
+    if (
+      options.parent &&
+      typeof (options.parent as types.Span).context === 'function'
+    ) {
+      parentSpanContext = (options.parent as types.Span).context();
+    }
+
+    // make sampling decision
+    if (!this.sampler.shouldSample(parentSpanContext)) {
+      return BasicTracer.defaultSpan;
+    }
+
+    // span context
+    const traceId = parentSpanContext
+      ? parentSpanContext.traceId
+      : randomTraceId();
+    const spanId = randomSpanId();
+
+    // TODO: create a real Span
+    const span = new NoopSpan({
+      traceId,
+      spanId,
+    });
+
+    // Set default attributes
+    span.setAttributes(this.defaultAttributes);
+
+    return span;
+  }
+
+  /**
+   * Returns the current Span from the current context.
+   */
+  getCurrentSpan(): types.Span {
+    // Return with defaultSpan if no scope manager provided.
+    if (!this.scopeManager) {
+      this.logger.warn(
+        'getCurrentSpan() returns an invalid default span without a scopeManager'
+      );
+      return BasicTracer.defaultSpan;
```

Comment:
I think this should be `null`. In the noop it makes sense to always return a span so that nothing breaks when you disable the tracer, but for the base implementation `null` makes more sense to indicate the absence of a span on the scope.

Reply:
See my previous comment (https://github.com/open-telemetry/opentelemetry-js/pull/99#discussion_r303208415) about the default scope manager. 
Looks like there are two opinions here: return null, log warning. Both can work for `getCurrentSpan()` and although I like the idea of `null` to indicate you really have no ScopeManager, the same strict check is not applicable for `withSpan` so I'd vote for warning log. I guess throwing in `withSpan` without ScopeManager would be quite abusive.

Reply:
Added to SIG agenda.

Reply:
Actually i think it's also applicable to `withSpan` when using the `NoopScopeManager`, it will just always return a `null` active scope, which is fine (at least for me).

Reply:
A few thoughts:

* The TS type should include `null` anyway as it is not guaranteed that there is always an active span.
* The no-op **must** return a span always, otherwise disabling the tracer could break any code that doesn't check for `null`.
* The default behavior to always return `null` makes the most sense since there is no scope manager. If you don't want scope management, then that's what you expect. If you want scope management, then a scope manager should be added. The default behavior should not log warnings, otherwise it means the default is wrong. FWIW we have users that explicitly don't want scope management, and they definitely don't want a warning because of it.

Reply:
> we have users that explicitly don't want scope management, and they definitely don't want a warning because of it.

Do you know what would be a use-case that you don't want scope management but you still want to call `withSpan` and `getCurrentSpan` however you know they don't do anything?

Reply:
(the default null without warning for `getCurrentSpan` would make sense for me as it shows well that you cannot use this feature without scope manager, although the `withSpan` would just silently swallow this fact, I don't think you can enforce it on API level, except if you throw)

Reply:
> Do you know what would be a use-case that you don't want scope management but you still want to call withSpan and getCurrentSpan however you know they don't do anything?

The auto instrumentation will call these methods.

Reply:
Isn't ScopeManager a hard requirement for auto instrumentation? Otherwise, you don't have a root span. How would it work?

Reply:
That's a good point. Hopefully it will be possible to have at least some level of auto instrumentation even without a scope manager, but it would definitely be challenging. How would the auto instrumentation detect the presence of a proper scope manager?